vault: Refactor ACL Creation Concurrency Test

alrs opened the pull request:
This PR fixes an error lost inside a goroutine in a test in the `vault` package.

`T.Fatal()` is a convenience function that combines `T.Log()` and `T.FailNow()`. 

From the godoc of `T.FailNow()`: 

> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines.

To see this in action, take a look at `main_test.go` in my [example repo](https://github.com/alrs/fatalfail/).

ncabatoff commented:
Hi @alrs, thank you for the contribution.  I didn't realize that about FailNow, good to know.  I think there are a few problems with your solution.  First, you only consume a single value from the channel, but there may be many written, so you could miss an err.  There may also be zero written, in which case your read from the channel will block until the test times out.  Finally, by eliminating the waitgroup we no longer have any synchronization between goroutine completion and evaluating the result.

I suggest you have a look at https://github.com/oklog/run, I think it may work well here.

alrs commented:
@ncabatoff as this anonymous goroutine is written here, only one variable can be sent down the channel. If the timeout is hit, a `nil` goes down `errCh` and the goroutine returns. If an error is hit, it goes down the `errCh` and the goroutine returns.

Synchronization is maintained, because the `err = <- errCh` on line 823 blocks until either an error or a nil come down `errCh`.

I put an example on the Go playground: https://play.golang.org/p/XkBHjNuuhKm

ncabatoff commented:
There is one channel, and 50 goroutines, and if they all fail 50 errors will be sent.  It is entirely possible for neither a timeout nor an error to occur, in which case nothing is sent on the channel.

alrs commented:
@ncabatoff You're right, I had my eye on the loop on 809, not 807.

calvn commented on vault/acl_test.go:
```diff
@@ -801,26 +801,29 @@ func TestACL_CreationRace(t *testing.T) {
 				}
 			}
 		}()
+		err = <-errCh
```
This would block the for loop until the error is received before the next goroutine gets called.

alrs commented:
@ncabatoff I thought I had a clever solution to stick the `errCh` inside the loop, but that obviously kills the parallelism. I'll mark this PR as WIP and make some changes.

calvn commented:
I think we could simply swap `t.Fatalf` with `t.Errorf`. This way, the errors gets logged and the test is marked a failed if so, but execution continues all the way through which avoids failing fast and swallowing other possible errors.

ncabatoff commented:
Correction to my earlier comments: I failed to notice the inner `for{}` loop, so my claims that it's possible for a goroutine to yield 0 writes to the channel were false.

alrs commented:
@calvn I don't believe that `t.Fail()` is of any help here.

calvn commented:
Is the issue here not being able to log all errors from the 50 goroutines due to the nature of `t.Fatal` if one of them fails first? Since `t.Fatal` is called within a goroutine, we actually don't break out of the test case early.

alrs commented:
@calvn once a `testing.T` variable goes into a goroutine it loses the ability to log or fail.

I'm potentially collecting more than one error here and throwing the rest away, but I'm using the close of the `errCh` as my means of blocking until tests are done in lieu of `sync.WaitGroup`.

I like how this turned out: got rid of some magic numbers, using a `select` to wait for either an error or a timeout, using the goroutine-safe `sync/atomic` package to keep track of the goroutines.

Thanks @ncabatoff for your insights.

calvn commented:
>  once a testing.T variable goes into a goroutine it loses the ability to log or fail.

I don't think that's the case. Calling `FailNow()` within a goroutine should still mark the test as failed. The `t.Fatalf` call in this case should therefore still fail the test case and log the errors. It wouldn't fail the parent test goroutine immediately, though.

See https://play.golang.org/p/XDBVcpdoH-X.

ncabatoff commented:
I suggested oklog's run groups, but what I should've said was golang.org/x/sync/errgroup.  That gives a nice succinct equivalent to what we had before:

```go
	var group errgroup.Group
	stopTime := time.Now().Add(20 * time.Second)

	for i := 0; i < 50; i++ {
		group.Go(func() error {
			for {
				if time.Now().After(stopTime) {
					return nil
				}
				_, err := NewACL(namespace.RootContext(nil), []*Policy{policy})
				if err != nil {
					return err
				}
			}
		})
	}

	err = group.Wait()
	if err != nil {
		t.Fatal(err)
	}
```

alrs commented:
Thanks, @calvn. You're right.

My WAG is that this is an improvement in Go that showed up in a recent release, but I don't have access to any old godoc handy to compare.

The comment on 798 is `// NOTE: this test doesn't catch any races ATM`, which leads me to suspect the `testing.T`-in-a-goroutine story used to be worse in the past than it is today.

I'm still happy with this PR, if you guys are. I'm happy to close it if you want to go another direction.

ncabatoff commented:
My interpretation of the comment on 798 is they're saying that although the test is designed to try to provoke failures due to races, it is unable to (which is a good thing).

I think the concern you raise about Fatal/FailNow in goroutines is still valid, it's documented in the testing package:

```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. 
```

It is ambiguous from that comment what the impact of disregarding it would be.  Clearly it works sometimes, but I'm hesitant to assume it's safe without studying the testing code.

calvn commented:
The intent of `t.Fatalf` (i.e. failing fast) is not immediately obvious if it's called within a goroutine since `FailNow`s behavior no longer applies. From a readability perspective, changing it to `t.Errorf` would be more appropriate.

ncabatoff commented:
Sure, but typically we prefer to fail fast, which is why I advocate for the errgroup approach.  You get the waitgroup semantics, plus as soon as an error occurs it bubbles up and stops everything.

calvn commented:
Agreed. I wasn't sure if the intent of this PR was to capture all dropped errors from the goroutines, or whether we wanted to simply capture the first error and fail fast. If it's the latter, errgroup is definitely the way to go.

alrs commented:
@ncabatoff @calvn I switched over to using an error group, and left the changed test logic.

Previously, there was a `for` loop that set loose 50 goroutines that called `NewACL()` as many times as they could within 20 seconds.

That seems like a recipe for flakes to me, since the number of goroutines would vary run-to-run and machine-to-machine.

Instead, I upped the number of times the `for` loop executes significantly, and treat 20 seconds as a deadline instead of the default run time.

jefferai commented:
> That seems like a recipe for flakes to me, since the number of goroutines would vary run-to-run and machine-to-machine.

I think that was the point. Different machines will display different behaviors, which gives tests different conditions under which they might fail or e.g. trigger the race detector.

alrs commented:
@jefferai There is still a lot of room for variation between architectures, core-counts, and go versions in firing off 100k goroutines.

CI environments are usually the most constrained, so having tests that are less likely to fail there leads to people defaulting to ignoring local test runs.

While you're here, do you think it's reasonable to remove the `//NOTE: this test doesn't catch any races ATM` from line 797?

alrs commented:
Thanks, everyone for tolerating my circleci spam. I can pretty reliably hit 15k concurrent ACL creations on my local laptop, but only 500 on circleci.

I changed up the test again. I restored the "fire-off-as-many-goroutines-as-you-can-in-20-seconds" functionality, but now limit how many goroutines can run simultaneously. I also added an error condition for a situation where the required concurrency couldn't be accomplished.

As the test was, the absolute worst-case scenario was that `NewACL()` would never happen at all, because the first goroutine didn't manage to happen within the required 20 seconds. 

As the test is, a lack of meaningful concurrency causes a failure and an explanation as to why.

alrs commented:
@ncabatoff @calvn are you two on-board with this?

ncabatoff commented:
I have misgivings about the new approach.  I get where you're coming from: what's the use of having a test run which doesn't achieve its purpose because it was under-provisioned?  But the flip side of that is that the test isn't deterministic anyway, and in this case I would rather have false negatives (where the test didn't detect a race because it didn't have enough CPU) than have tests fail without there being an actual problem we can remedy.  In other words if we're unlucky with the circleci env we get, the test will fail, and a developer will have to take time to investigate it - or will have trained themself to ignore this particular failure as a false positive.

In short: I'd rather the test be less effective than have it fail intermittently with no recourse.   

We have some tests that don't run in circleci because they rely on something we don't provide the circleci env, e.g. cloud credentials.  Maybe we should be treating this test similarly, as something to be run periodically in a more controlled environment.

alrs commented:
@ncabatoff I changed the test to `T.Skip()` in circleci, and cranked the concurrency back up to 10k, which runs and stresses my `Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz`.

This was a fun exercise for me. I'd be happy to see this merged, and I'd be just as happy to close this PR and leave this test be.

alrs commented:
OK, another swing at improving this test.

I changed `TestACL_CreationRace()` to `T.Skip()` if desired concurrency isn't reached. This way you get a message saying that the test didn't really count, but it doesn't cause failure in weak test environments. This gets rid of the code I had that read environment variables to see if the test was happening in CircleCI.

I like that this PR has an upper-bound of concurrency, so that

VioletHynes commented:
As this PR is older and cannot currently be merged due to merge conflicts, I’m going to go ahead and close it. We've since updated this test, and I don't think it's fair to ask you to resolve all of the merge conflicts for a four year old test PR. Do feel free to submit new PRs in the future!
